### PR TITLE
Require python 3.9, plus lots of linting and formatting fixes

### PR DESCRIPTION
**Pull request by timwoj**
I added a basic pyproject.toml to this just for consistency with our other projects. I didn't add a setup.py because you will generally never install this package outside of a Zeek installation.

### CODE REVIEW

File: ZeekControl/web.py [hunk 1]
```diff
@@ -1,59 +1,64 @@
...
-    return {"id": i} 
+    return {"id": i}
+
 
 def run_app(client):
     app.daemon = client
-    bottle.run(app, host='localhost', port=8082)
-
-if __name__ == "__main__":
-    main()
```
Comment (bbannier) [inline ZeekControl/web.py:59]: Uff.

File: testing/Cfg/bin/zeek__test [hunk 2]
```diff
@@ -96,13 +107,16 @@ def readzeekctltestcfg():
...
 
-    with open(cfgfilepath, "r") as fin:
+    with open(cfgfilepath) as fin:
         for line in fin:
-            key, val = line.strip().split('=', 1)
+            key, val = line.strip().split("=", 1)
             key = key.strip()
             val = val.strip().split()
             if key not in keys:
-                print("Error: unknown option '%s' in: %s" % (zeekvar, cfgfilepath), file=sys.stderr)
```
Comment (bbannier) [inline testing/Cfg/bin/zeek__test:105]: Uff.

File: bin/stats-to-csv [hunk 3]
```diff
@@ -47,18 +46,16 @@ def readNodes(meta):
...
-
-        name = os.path.join(wwwdir, "%s.%s.csv" % (node, tag))
+        name = os.path.join(wwwdir, f"{node}.{tag}.csv")
 
         if os.path.exists(name):
             f = open(name, "a")
         else:
             f = open(name, "w")
-            f.write("time,%s\n" % ",".join(columns))
+            f.write("time,{:s}\n".format(",".join(columns)))
```
Comment (bbannier) [inline bin/stats-to-csv:58]: String formatting is the default.

```suggestion
            f.write("time,{}\n".format(",".join(columns)))
```

File: testing/Scripts/diff-to-bytes-output [hunk 4]
```diff
@@ -18,8 +17,8 @@ for line in sys.stdin:
             num = 0
 
         if num > bignumber:
-            outstr += "%.5e " % num
+            outstr += f"{num:.5e} "
         else:
-            outstr += "%s " % field
+            outstr += f"{field} " % field
```
Comment (bbannier) [inline testing/Scripts/diff-to-bytes-output:22]: Stray arg.

```suggestion
            outstr += f"{field} "
```

File: .pre-commit-config.yaml
Comment (bbannier) [inline .pre-commit-config.yaml:1]: Given how much we flagged here let's please add a GH job running pre-commit.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    exclude: ^testing/Baseline
+  - id: end-of-file-fixer
+    exclude: ^testing/Baseline|examples/.*Baseline.*
+
+- repo: https://github.com/astral-sh/ruff-pre-commit
+  rev: v0.8.1
+  hooks:
+    - id: ruff-format
+    - id: ruff
+      args: [--fix]
```
Comment (bbannier) [inline .pre-commit-config.yaml:17]: Nit: inconsistent formatting.

```suggestion
  - id: ruff-format
  - id: ruff
    args: [--fix]
```

File: .github/workflows/codeql.yml
Comment (bbannier) [inline .github/workflows/codeql.yml:1]: Tangentially related, most actions here are outdated and we should bump them, e.g.,

```diff
diff --git a/.github/workflows/codeql.yml b/.github/workflows/codeql.yml
index 95e91c6..b01ff18 100644
--- a/.github/workflows/codeql.yml
+++ b/.github/workflows/codeql.yml
@@ -24,10 +24,10 @@ jobs:
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@v3
+      uses: actions/checkout@v4
 
     - name: Initialize CodeQL
-      uses: github/codeql-action/init@v2
+      uses: github/codeql-action/init@v3
       with:
         languages: ${{ matrix.language }}
         # If you wish to specify custom queries, you can do so here or in a config file.
@@ -39,9 +39,9 @@ jobs:
 
 
     - name: Autobuild
-      uses: github/codeql-action/autobuild@v2
+      uses: github/codeql-action/autobuild@v3
 
     - name: Perform CodeQL Analysis
-      uses: github/codeql-action/analyze@v2
+      uses: github/codeql-action/analyze@v3
       with:
         category: "/language:${{matrix.language}}"
```

If we want we could lint this automatically from pre-commit by adding https://github.com/rhysd/actionlint as a hook, e.g.,

```yaml
- repo: https://github.com/rhysd/actionlint
  rev: v1.7.4
  hooks:
  - id: actionlint
```

**Review by bbannier — APPROVED** [review]
This is a huge PR 😬. I tried to review individual commits, but not all pass the pre-commit hooks individually. What I did instead was at the tip revert all non-linter changes to their version from `master`, automatically apply all changes and review the delta.

The changes look fine and mostly cosmetic to me, but you also fixed a number of true bugs (noted them with _Uff_ comments in the review). I did not test how they changed behavior, but they looked reasonable to me.

Let's definitely add a GH job running pre-commit. It would also be great to have some CI running the test suite for the things we cover, but that might be hard. As is this package cannot be installed as a Python project.

File: pyproject.toml
[the remark this answers is not in the document]
Comment (timwoj) [reply pyproject.toml:1]: Noted. I removed that change. Talking to @ckreibich we never intended for this to be installed as a separate package outside of a Zeek installation, so all of the various project setup files were never created.

File: .github/workflows/codeql.yml
Comment (timwoj) [reply .github/workflows/codeql.yml:1]: Fixed the versions. The `actionlint` thing looks useful across all of the Zeek projects too.

**Pull request merged**